### PR TITLE
040 login system

### PR DESCRIPTION
**Pull request by MylesScullion**
Login system for the frontend

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+app.set('view engine', 'html');
+
+app.use('/public', express.static(path.join(__dirname, 'public')));
+
+app.use(express.json())
+
+
+app.use(express.urlencoded({ extended: true }));
+
+app.use(session({secret: 'NOT HARDCODED SECRET', cookie: {maxAge: 60000}}));
```
Comment (glenn-hamilton) [inline app.ts]: Your secret shouldn't be hard coded, it should be passed via an environment variable

File: controller/authController.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        res.render('login')
+    })
+
+    app.post('/login', async (req: Request, res: Response)=> {
+        let data: Login = req.body
+
+        try{
+            const jwtToken = await authService.login(data);
+
+            res.cookie('jwtToken', jwtToken, { maxAge: 3600000, httpOnly: true });
```
Comment (glenn-hamilton) [inline controller/authController.ts]: How'd you come to this `maxAge` value?

File: model/employees.ts
Comment (glenn-hamilton) [inline model/employees.ts:1]: This file is empty

File: package.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "belfast-team2-frontend",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
Comment (glenn-hamilton) [inline package.json]: Should this be `app.js`?

File: service/authService.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import {Login, ActiveSession, User } from "../model/auth";
+const axios =require('axios');
+
+module.exports.login = async function (login: Login): Promise<ActiveSession>{
+    try{
+        const response= await axios.post('http://localhost:8080/api/login', login)
+        
+        return response.data
+    }catch (e) {
+        // if(e.response.status === 401){
```
Comment (glenn-hamilton) [inline service/authService.ts]: Is this still needed?

File: service/authService.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import {Login, ActiveSession, User } from "../model/auth";
+const axios =require('axios');
+
+module.exports.login = async function (login: Login): Promise<ActiveSession>{
+    try{
+        const response= await axios.post('http://localhost:8080/api/login', login)
```
Comment (glenn-hamilton) [inline service/authService.ts]: This hard coded URL won't work correctly when deployed to AWS

File: service/authService.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        //     throw new Error ('Your Email or Password is incorrect') 
+        // }
+        throw new Error ('Unable to Login')
+       
+    }
+}
+
+module.exports.register = async function (user: User): Promise<void>{
+    try{
+        const response= await axios.post('http://localhost:8080/api/register', user)
```
Comment (glenn-hamilton) [inline service/authService.ts]: Same with this URL

File: controller/authController.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+            res.render('login', req.body)
+        }
+    })
+
+    app.get('/register', async (req: Request, res: Response)=>{
+        res.render('register')
+    })
+
+    app.post('/register', async (req: Request, res: Response)=> {
```
Comment (glenn-hamilton) [inline controller/authController.ts:35]: There is no positive feedback to the user here to let them know that their account has been successfully created

File: service/authService.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        
+        return response.data
+    }catch (e) {
+        throw new Error ('Registration failure')
```
Comment (glenn-hamilton) [inline service/authService.ts]: This error isn't very useful to the user

File: service/authService.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import {Login, ActiveSession, User } from "../model/auth";
+const axios =require('axios');
+require('dotenv').config();
+
+module.exports.login = async function (login: Login): Promise<ActiveSession>{
+    try{
+        const response= await axios.post(process.env.URL+'/login/'+login)
```
Comment (shaunganley) [inline service/authService.ts]:
```suggestion
        const response= await axios.post(process.env.URL+'/login/', login)
```

File: service/authService.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    }catch (e) {
+       
+        throw new Error ('Unable to Login')
+       
+    }
+}
+
+module.exports.register = async function (user: User): Promise<void>{
+    try{
+        const response= await axios.post(process.env.URL+'/register/'+user)
```
Comment (shaunganley) [inline service/authService.ts]:
```suggestion
        const response= await axios.post(process.env.URL+'/register/', user)
```

**Review by shaunganley — CHANGES_REQUESTED** [review]
really impressive stuff. Just 2 comments on how you're passing the request body using axios. You will also need to add some unit and UI tests

File: Validator/RegisterValidatorTest.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import { Register } from "../model/auth";
+
+module.exports.validateRegister = function (register: Register): string {
```
Comment (shaunganley) [inline Validator/RegisterValidatorTest.ts]: this is identical to validateLogin method

File: controller/authController.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        let data: Register = req.body
+
+        try{
+            await authService.register(data)
+
+            res.redirect('/login?success_msg=You are now registered and can log in');
```
Comment (shaunganley) [inline controller/authController.ts]: you're not actually displaying this on the login page

**Review by shaunganley — CHANGES_REQUESTED** [review]
still needs unit tests for service

File: .gitignore [hunk 13]
```diff
@@ -1,2 +1,5 @@
 node_modules/
-dist/
\ No newline at end of file
+dist/
+.env/
+.env
```
Comment (shaunganley) [inline .gitignore]: duplicate

File: Validator/Validator.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        if (!register.email.includes("@")) {
+            return "Email must contain an '@' symbol";
+        }
+
+        if (register.password.length < 8 || register.password.length > 64) {
+            return "Password length not between 8 and 64 characters";
+        }
+
+        return null;
+    }
```
Comment (shaunganley) [inline Validator/Validator.ts]: duplicated code

File: test/service/ValidateServiceTest.js [hunk 15]
```diff
@@ -0,0 +1,84 @@
+const { mock } = require('jest-mock-extended');
+const Validator = require('../path/to/validator');
+const Login = require('../path/to/login');
+const Register = require('../path/to/register');
+
+describe('ValidatorServiceTest', () => {
+    let validator;
+
+    beforeEach(() => {
+        validator = mock(Validator);
```
Comment (shaunganley) [inline test/service/ValidateServiceTest.js]: don't mock the validator in the validator test class

File: test/unit/validator/LoginValidatorTest.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
+const { validateLogin } = require('./validateLogin');
+const { Login } = require('../model/auth'); 
+
+describe('validateLogin', () => {
+    test('isValidLogin_EmailTooLong', () => {
```
Comment (shaunganley) [inline test/unit/validator/LoginValidatorTest.js]: can be removed

File: test/unit/validator/RegisterValidatorTest.js [hunk 17]
```diff
@@ -0,0 +1,28 @@
+const { isValidRegister } = require('./isValidRegister');
+const { Register } = require('../model/auth');
+
+describe('isValidRegister', () => {
```
Comment (shaunganley) [inline test/unit/validator/RegisterValidatorTest.js]: can be removed

**Review by shaunganley — CHANGES_REQUESTED** [review]
still missing tests for authService

File: test/unit/service/authServiceTest.ts [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        mock.restore();
+    });
+
+    it('should login successfully', async () => {
+        const loginData = { email: "test@example.com", password: "password123" };
+    
+        mock.onPost(process.env.URL+'/login/', loginData).reply(200);
+    
+        try {
+            await Login(loginData);
```
Comment (shaunganley) [inline test/unit/service/authServiceTest.ts]: you're not calling your service method here to actually test it

File: test/unit/service/authServiceTest.ts [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    
+    });
+    it('should throw error when login fails', async () => {
+        const loginData = { email: "test@example.com", password: "password123" };
+    
+        mock.onPost(process.env.URL+'/login/', loginData).reply(500);
+    
+        try {
+            await Login(loginData);
```
Comment (shaunganley) [inline test/unit/service/authServiceTest.ts]: you're not calling your service method here to actually test it

File: test/unit/service/authServiceTest.ts [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    });
+    
+    it('should register successfully', async () => {
+        const registerData = { email: "test@example.com", password: "password123", role: "Admin" };
+    
+        mock.onPost(process.env.URL+'/register/', registerData).reply(200);
+    
+        try {
+            await Register(registerData);
```
Comment (shaunganley) [inline test/unit/service/authServiceTest.ts]: you're not calling your service method here to actually test it

File: test/unit/service/authServiceTest.ts [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+        }
+    });
+    
+    it('should throw error when registration fails', async () => {
+        const registerData = { email: "test@example.com", password: "password123", role: "Admin" };
+    
+        mock.onPost(process.env.URL+'/register/', registerData).reply(500);
+    
+        try {
+            await Register(registerData);
```
Comment (shaunganley) [inline test/unit/service/authServiceTest.ts]: you're not calling your service method here to actually test it

**Review by shaunganley — CHANGES_REQUESTED** [review]

**Review by shaunganley — APPROVED** [review]

**Review by omcmenamin — APPROVED** [review]

**Pull request merged**